### PR TITLE
Add branch-based versioning

### LINKED ISSUE #3573: De-Risk hosting multiple versions of the Knative website
## Description: 
This is a story for de-risking multiple versions using MkDocs. **This story only requires that the previous versions build, not that they must be constantly up-to-date based on cherrypicked PRs to the respective `release-{version}` branches.** 

That functionality will be covered in a different story. 

---
## Story
```
As Daniela the Developer
I want to see previous versions of the Knative.dev website
So that I can see documentation relevant to the version I have deployed
```

## Acceptance Criteria
✅ `Version Dropdown`
```
GIVEN I am on the Knative.dev homepage (https://knative.github.io/docs/docs/)
WHEN I look at the header 
THEN I see a version dropdown with the "Development" version selected
(EDIT) AND I see a big black banner letting me know I am on an old, unsupported version
```

⚙️ `Multiple Versions Present`
```
GIVEN I am on the Knative.dev homepage (https://knative.github.io/docs/docs/)
WHEN I click on the Version Dropdown
THEN I see the last 4 releases (0.22-0.19), Archives, and Pre-Release as options 
AND Pre-Release is greyed out
```
^ ⚙️ Missing N-4 versions (0.19)

⚙️`Visit Older Versions`
```
GIVEN I have clicked on the Version Dropdown 
WHEN I click on an older version (i.e. 0.19) 
THEN I see the content which lives on the release-{version} (i.e. 0.19) branch in Github 
AND it renders in the old Hugo format 
```
^⚙️  Not rendered in old Hugo format, switched to MkDocs rendering 

## Known Issues and Considerations: 
- [x] For this implementation, we do not need to support cherrypicking of old PRs, we will cover that in a later issue.

### PR DESCRIPTION
**Pull request by julz**
https://user-images.githubusercontent.com/354013/118398771-073ddb80-b652-11eb-90d9-24dcca25f255.mov

Builds previous versions of docs in to the site based on release branches.

**TL;dr:** `mkdocs serve` in any branch for live preview of just that version of docs, `./hack/build-with-blog.sh [serve]` to build/preview the whole thing, release by creating release branch as normal and updating RELEASES at top of build-with-blog.sh.

Fixes #3573.

/assign @omerbensaadon @csantanapr 

# Preview docs with live reload

- `mkdocs serve --dirtyreload` - this will preview just the current version, very fast (immediate) update of changes to pages
- Run on a release branch (or on main, or wherever) to view docs for just that release.

# Preview versioned docs, blog posts and `main` of community

- `./hack/build-with-blog.sh serve`
- We'll set netlify PR previews (and builds) up to run this, too.

# Preview community repo PRs

- `COMMUNITY_BRANCH=${pr branch here} COMMUNITY_REPO=${pr author username} ./hack/build-with-blog.sh serve`. This will build community tab from the given branch of the given repo instead of main of knative/docs, good for previewing PRs to community before merging.

# Build full versioned docs with blog and community included

- `./hack/build-with-blog.sh` - builds the whole shebang. HEAD -> /development, latest release -> /docs etc. This is what actions/netlify builds will run.
- Automatically adds version warning banners to old pages, sets correct version for artifact links etc
- Note: this is currently using my julz/docs repo's mkrelease-0.23 and mkrelease-0.22 branches until we have past version branches created on the main repo. This can be changed via the BRANCHES array at the top of the script.

# Releasing

- Create release branch, just like today
- Update RELEASES array at top of hack/build-with-blog.sh (on main/mkdocs) to tell us what releases to build. PR the result.
- That's it.

### CODE REVIEW

File: hack/build-with-blog.sh [hunk 1]
```diff
@@ -1,53 +1,111 @@
...
+  # See https://github.com/knative/website/blob/main/scripts/processsourcefiles.sh#L125
+  # For the reasoning behind all this.
+  echo 'Converting all links in GitHub source files to Hugo supported relative links...'
+  pushd temp/website
+  # Convert relative links to support Hugo
+  find . -type f -path '*/content/*.md' ! -name '*_index.md' ! -name '*index.md' ! -name '*README.md' \
+    ! -name '*serving-api.md' ! -name '*eventing-contrib-api.md' ! -name '*eventing-api.md' \
+    ! -name '*build-api.md' ! -name '*.git*' ! -path '*/.github/*' ! -path '*/hack/*' \
+    ! -path '*/node_modules/*' ! -path '*/test/*' ! -path '*/themes/*' ! -path '*/vendor/*' \
+    -exec sed -i '/](/ { s#(\.\.\/#(../../#g; s#(\.\/#(../#g; }' {} +
```
Comment (csantanapr) [inline hack/build-with-blog.sh:92]: I guess we need the special  GNU `sed` on MacOS, same way as we have today in website repo build
I got an error
```
+ find . -type f -path '*/content/*.md' '!' -name '*_index.md' '!' -name '*index.md' '!' -name '*README.md' '!' -name '*serving-api.md' '!' -name '*eventing-contrib-api.md' '!' -name '*eventing-api.md' '!' -name '*build-api.md' '!' -name '*.git*' '!' -path '*/.github/*' '!' -path '*/hack/*' '!' -path '*/node_modules/*' '!' -path '*/test/*' '!' -path '*/themes/*' '!' -path '*/vendor/*' -exec sed -i '/](/ { s#(\.\.\/#(../../#g; s#(\.\/#(../#g; }' '{}' +
sed: 1: "./content/en/blog/relea ...": invalid command code .
```

Comment (julz) [reply hack/build-with-blog.sh:92]: yeah, this is a copy paste from website repo (there's a couple links in community that need this while we're still building it with hugo), so same prereqs from there will be needed for this :(

Comment (csantanapr) [reply hack/build-with-blog.sh:92]: Can we add this to the script, and later in the MKDOCS.md dev guide @omerbensaadon is working on.

```
# 2. For Mac OSX: The script uses the `gnu` version of `sed`. To install `gnu-sed`, you use brew:#    1. Run `brew install gnu-sed`#    2. Add it to your `PATH`. For example, add the following line to your `~/.bash_profile`:#      `PATH="/usr/local/opt/gnu-sed/libexec/gnubin:$PATH"`
```
This is the same we have in website repo https://github.com/knative/website/blob/9e5693d8cdf47af18ac1587be601b9e6c610d321/scripts/localbuild.sh#L13

Comment (julz) [reply hack/build-with-blog.sh:92]: Might be able to do this in a cross-platform way (it is possible, I think, dont know why we didnt already), but I think we should probably land that separately

File: overrides/main.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{% extends "base.html" %}
+
+{% block announce %}
+{% if config.extra.version_warning %}
+<div class="versionwarning" style="margin: .6rem auto; padding: 0 .8rem">
+  <h1 style="color: #ffcc00">You are viewing documentation for Knative version: {{ config.extra.knative_version }}</h1>
+  <p>
+  Knative {{ config.extra.knative_version }} documentation is no longer actively maintained. The version you are currently viewing is a static snapshot. For up-to-date documentation, see <a href="http://knative.dev/docs">the latest version</a>.
```
Comment (csantanapr) [inline overrides/main.html:8]: Is better to make this relative `href="http://knative.dev/docs"`
This way it works from github pages, or locally

**Review by csantanapr — CHANGES_REQUESTED** [review]

File: hack/build-with-blog.sh [hunk 3]
```diff
@@ -1,53 +1,111 @@
...
+  # Convert all relative links from README.md to index.html
+  find . -type f -path '*/content/*.md' ! -name '_index.md' \
+      -exec sed -i '/](/ { /http/ !{s#README\.md#index.html#g} }' {} +
+  # Convert all Markdown links to HTML
+  find . -type f -path '*/content/*.md' \
+      -exec sed -i '/](/ { /http/ !{s#\.md##g} }' {} +
+
+  # Run the hugo build as normal!
+  npx hugo
```
Comment (csantanapr) [inline hack/build-with-blog.sh:101]: This is failing for me
```
+ npx hugo
npm ERR! could not determine executable to run
```

And I have hugo installed, 
```
which hugo
/usr/local/bin/hugo
```
I had impression `npx` takes care of installing, No?

File: hack/build-with-blog.sh (hunk 1 above)
Comment (csantanapr) [reply hack/build-with-blog.sh:92]: yeah that sounds, good create an issue label mkdocs for the backlog to simply sed. for now just add the instructions in the script for folks to install with brew

Comment (julz) [reply hack/build-with-blog.sh:99]: added a bit of docs to the top of the script for now, hopefully we can also create a docker container so folks dont have to think about this too much

File: hack/build-with-blog.sh (hunk 3 above)
Comment (julz) [reply hack/build-with-blog.sh:108]: I think it just runs it with the right PATH set, you still have to do `npm install`.. don't know why that isn't working for you tho 🤔🤔

Comment (csantanapr) [inline hack/build-with-blog.sh:108]:
```suggestion
hugo
```
I changed but know postcss not working
```
+ hugo
Start building sites … 
Total in 707 ms
Error: Error building site: POSTCSS: failed to transform "scss/main.css" (text/css). Check your PostCSS installation; install with "npm install postcss-cli". See https://gohugo.io/hugo-pipes/postcss/
```
maybe we should add to the path `export PATH="$PWD/node_modules/.bin:$PATH"`

Comment (csantanapr) [reply hack/build-with-blog.sh:108]: oh yes that's what we already have
```
 Run the hugo build as normal!

# need postcss cli in PATH
PATH=${PATH}:${PWD}/node_modules/.bin
pushd temp/website
hugo
popd
```

File: overrides/main.html (hunk 2 above)
Comment (julz) [reply overrides/main.html:8]: it's a little tricky to do it relatively since the include is static but could be added to any pages. I might need to think about this one a bit (it's probably an easy solution Im just not thinking of).

File: hack/build-with-blog.sh (hunk 3 above)
Comment (csantanapr) [reply hack/build-with-blog.sh:108]: yep with worked with 
```
# need postcss cli in PATH
PATH=${PATH}:${PWD}/node_modules/.bin
hugo
popd
```

Comment (julz) [reply hack/build-with-blog.sh:108]: I think `export PATH="$PWD/node_modules/.bin:$PATH"` is exactly what npx does...

Comment (julz) [reply hack/build-with-blog.sh:108]: hmm hmm hmm wonder why npx doesnt work but this does 🤔 maybe my mental model of what npx does is wrong

Comment (julz) [reply hack/build-with-blog.sh:108]: doesnt really matter, Ill just do the PATH hack for now

File: overrides/main.html (hunk 2 above)
Comment (csantanapr) [reply overrides/main.html:8]: I think using `href="../docs` should work and cover HTTPS

File: hack/build-with-blog.sh (hunk 1 above)
Comment (csantanapr) [reply hack/build-with-blog.sh:99]: We already have a docker container @julz or you talking about something else?
https://github.com/knative/docs/blob/mkdocs/hack/docker/README.md

File: hack/build-with-blog.sh (hunk 3 above)
Comment (csantanapr) [reply hack/build-with-blog.sh:108]: maybe you were thinking on implementing `npm run hugo` and have a script entry in package.json?

File: overrides/main.html (hunk 2 above)
Comment (csantanapr) [reply overrides/main.html:8]: This works for me
image

But the HTML generated should preserve it to `../docs` and not `knative.dev/docs`

Comment (csantanapr) [reply overrides/main.html:8]: hmm weird using `../docs` generates `<a href="http://knative.dev/docs">`
image

How can we make it stick to `../docs`

Comment (csantanapr) [reply overrides/main.html:8]: I guess make it `../docs` and we can fix later maybe with some javascript

File: hack/build-with-blog.sh (hunk 1 above)
Comment (julz) [reply hack/build-with-blog.sh:99]: I mean when we update it for this the install instructions should just be "run the container" iyswim

File: overrides/main.html (hunk 2 above)
Comment (julz) [reply overrides/main.html:8]: I dont think ../docs will work because you might be on /docs/getting-started/foo/bar/baz in which case that doesnt go to the right place

File: hack/build-with-blog.sh (hunk 3 above)
Comment (julz) [reply hack/build-with-blog.sh]: 🤷 🤷 🤷  no idea, but I pushed a version with the PATH set in an env var instead

**Comment by omerbensaadon** [comment]
once @csantanapr lgtm's I will approve and push :)

File: overrides/main.html (hunk 2 above)
Comment (julz) [reply overrides/main.html]: ok, it's a bit of a hack, but pushed something that I think will work

File: hack/build-with-blog.sh [hunk 4]
```diff
@@ -105,7 +105,7 @@ if [ -z "$SKIP_BLOG" ]; then
       -exec sed -i '/](/ { /http/ !{s#\.md##g} }' {} +
 
   # Run the hugo build as normal!
-  npx hugo
+  PATH=${PATH}:${PWD}/node_modules/.bin hugo
```
Comment (csantanapr) [inline hack/build-with-blog.sh:108]: like this one better

File: overrides/main.html [hunk 5]
```diff
@@ -5,7 +5,7 @@
 <div class="versionwarning" style="margin: .6rem auto; padding: 0 .8rem">
   <h1 style="color: #ffcc00">You are viewing documentation for Knative version: {{ config.extra.knative_version }}</h1>
   <p>
-  Knative {{ config.extra.knative_version }} documentation is no longer actively maintained. The version you are currently viewing is a static snapshot. For up-to-date documentation, see <a href="http://knative.dev/docs">the latest version</a>.
+  Knative {{ config.extra.knative_version }} documentation is no longer actively maintained. The version you are currently viewing is a static snapshot. For up-to-date documentation, see <a href="{{ base_url }}/../docs/">the latest version</a>.
```
Comment (csantanapr) [inline overrides/main.html:8]: nice trick 👍

**Review by csantanapr — CHANGES_REQUESTED** [review]
/lgtm
/approve

**Review by csantanapr — APPROVED** [review]
